### PR TITLE
conn: refine TLS configuration

### PR DESCRIPTION
**Pull request by gozssky**
### What problem does this PR solve? 
Fix CI for #1560. close https://github.com/pingcap/dm/pull/1560

### What is changed and how it works?
Set TLS too even if we only have ssl-ca.

### Check List 

Tests 

 - Unit test
 - Integration test
 - Manual test (add detailed scripts or steps below)
 
Related changes

 - Need to cherry-pick to the release branch

### CODE REVIEW

**Comment by lance6716** [comment]
```

[2021-04-09T07:49:19.975Z] PANIC: heartbeat_test.go:65: testHeartbeatSuite.TestHeartbeat

[2021-04-09T07:49:19.975Z] 

[2021-04-09T07:49:19.975Z] ... Panic: runtime error: invalid memory address or nil pointer dereference (PC=0x1439171)

[2021-04-09T07:49:19.975Z] 

[2021-04-09T07:49:19.975Z] /usr/local/go/src/runtime/panic.go:679

[2021-04-09T07:49:19.975Z]   in gopanic

[2021-04-09T07:49:19.975Z] /usr/local/go/src/runtime/panic.go:199

[2021-04-09T07:49:19.975Z]   in panicmem

[2021-04-09T07:49:19.975Z] /usr/local/go/src/runtime/signal_unix.go:394

[2021-04-09T07:49:19.975Z]   in sigpanic

[2021-04-09T07:49:19.975Z] heartbeat.go:126

[2021-04-09T07:49:19.975Z]   in Heartbeat.AddTask

[2021-04-09T07:49:19.975Z] heartbeat_test.go:73

[2021-04-09T07:49:19.975Z]   in testHeartbeatSuite.TestHeartbeat

[2021-04-09T07:49:19.975Z] /usr/local/go/src/reflect/value.go:321

[2021-04-09T07:49:19.975Z]   in Value.Call

[2021-04-09T07:49:19.975Z] /go/pkg/mod/github.com/pingcap/check@v0.0.0-20200212061837-5e12011dc712/check.go:850

[2021-04-09T07:49:19.975Z]   in suiteRunner.forkTest.func1

[2021-04-09T07:49:19.975Z] /go/pkg/mod/github.com/pingcap/check@v0.0.0-20200212061837-5e12011dc712/check.go:739

[2021-04-09T07:49:19.975Z]   in suiteRunner.forkCall.func1

[2021-04-09T07:49:19.975Z] /usr/local/go/src/runtime/asm_amd64.s:1357

[2021-04-09T07:49:19.975Z]   in goexit
``` @gozssky

**Comment by gozssky** [comment]
> ```
> 
> [2021-04-09T07:49:19.975Z] PANIC: heartbeat_test.go:65: testHeartbeatSuite.TestHeartbeat
> 
> [2021-04-09T07:49:19.975Z] 
> 
> [2021-04-09T07:49:19.975Z] ... Panic: runtime error: invalid memory address or nil pointer dereference (PC=0x1439171)
> 
> [2021-04-09T07:49:19.975Z] 
> 
> [2021-04-09T07:49:19.975Z] /usr/local/go/src/runtime/panic.go:679
> 
> [2021-04-09T07:49:19.975Z]   in gopanic
> 
> [2021-04-09T07:49:19.975Z] /usr/local/go/src/runtime/panic.go:199
> 
> [2021-04-09T07:49:19.975Z]   in panicmem
> 
> [2021-04-09T07:49:19.975Z] /usr/local/go/src/runtime/signal_unix.go:394
> 
> [2021-04-09T07:49:19.975Z]   in sigpanic
> 
> [2021-04-09T07:49:19.975Z] heartbeat.go:126
> 
> [2021-04-09T07:49:19.975Z]   in Heartbeat.AddTask
> 
> [2021-04-09T07:49:19.975Z] heartbeat_test.go:73
> 
> [2021-04-09T07:49:19.975Z]   in testHeartbeatSuite.TestHeartbeat
> 
> [2021-04-09T07:49:19.975Z] /usr/local/go/src/reflect/value.go:321
> 
> [2021-04-09T07:49:19.975Z]   in Value.Call
> 
> [2021-04-09T07:49:19.975Z] /go/pkg/mod/github.com/pingcap/check@v0.0.0-20200212061837-5e12011dc712/check.go:850
> 
> [2021-04-09T07:49:19.975Z]   in suiteRunner.forkTest.func1
> 
> [2021-04-09T07:49:19.975Z] /go/pkg/mod/github.com/pingcap/check@v0.0.0-20200212061837-5e12011dc712/check.go:739
> 
> [2021-04-09T07:49:19.975Z]   in suiteRunner.forkCall.func1
> 
> [2021-04-09T07:49:19.975Z] /usr/local/go/src/runtime/asm_amd64.s:1357
> 
> [2021-04-09T07:49:19.975Z]   in goexit
> ``` @gozssky 
> ```

I can be simply reproduced. I will check it.

File: syncer/heartbeat.go [hunk 1]
```diff
@@ -122,12 +123,12 @@ func (h *Heartbeat) AddTask(name string) error {
 	if h.primary == nil {
 		// open DB
 		dbCfg := h.cfg.primaryCfg
-		dbDSN := fmt.Sprintf("%s:%s@tcp(%s:%d)/?charset=utf8&interpolateParams=true&readTimeout=1m", dbCfg.User, dbCfg.Password, dbCfg.Host, dbCfg.Port)
-		primary, err := sql.Open("mysql", dbDSN)
+		dbCfg.RawDBCfg.ReadTimeout = "1m"
```
Comment (gozssky) [inline syncer/heartbeat.go:126]:
```suggestion
		if dbCfg.RawDBCfg != nil {
			dbCfg.RawDBCfg.SetReadTimeout("1m")
		}
```

**Review by gozssky — COMMENTED** [review]
@lance6716 PTAL

File: syncer/heartbeat.go (hunk 1 above)
Comment (lance6716) [reply syncer/heartbeat.go]: I think we could init a pointer to `RawDBCfg`

File: syncer/heartbeat.go [hunk 2]
```diff
@@ -122,12 +123,14 @@ func (h *Heartbeat) AddTask(name string) error {
 	if h.primary == nil {
 		// open DB
 		dbCfg := h.cfg.primaryCfg
-		dbDSN := fmt.Sprintf("%s:%s@tcp(%s:%d)/?charset=utf8&interpolateParams=true&readTimeout=1m", dbCfg.User, dbCfg.Password, dbCfg.Host, dbCfg.Port)
-		primary, err := sql.Open("mysql", dbDSN)
+		if dbCfg.RawDBCfg != nil {
+			dbCfg.RawDBCfg.SetReadTimeout("1m")
+		}
+		baseDB, err := conn.DefaultDBProvider.Apply(dbCfg)
```
Comment (gozssky) [inline syncer/heartbeat.go:129]:
```suggestion
		if dbCfg.RawDBCfg == nil {
			dbCfg.RawDBCfg = config.DefaultRawDBConfig()
		}
		dbCfg.RawDBCfg.ReadTimeout = "1m"
```

File: syncer/heartbeat.go (hunk 1 above)
Comment (gozssky) [reply syncer/heartbeat.go:129]: ok

**Review by lance6716 — COMMENTED** [review]
rest LGTM

File: relay/relay.go [hunk 3]
```diff
@@ -920,9 +919,8 @@ func (r *Relay) Reload(newCfg *Config) error {
 	r.cfg.Charset = newCfg.Charset
 
 	r.closeDB()
-	cfg := r.cfg.From
-	dbDSN := fmt.Sprintf("%s:%s@tcp(%s:%d)/?charset=utf8mb4&interpolateParams=true&readTimeout=%s", cfg.User, cfg.Password, cfg.Host, cfg.Port, showStatusConnectionTimeout)
-	db, err := sql.Open("mysql", dbDSN)
+	r.cfg.From.RawDBCfg.ReadTimeout = showStatusConnectionTimeout
```
[the remark this answers is not in the document]
Comment (gozssky) [reply relay/relay.go:925]: added

File: relay/relay.go [hunk 4]
```diff
@@ -920,9 +919,11 @@ func (r *Relay) Reload(newCfg *Config) error {
 	r.cfg.Charset = newCfg.Charset
 
 	r.closeDB()
-	cfg := r.cfg.From
-	dbDSN := fmt.Sprintf("%s:%s@tcp(%s:%d)/?charset=utf8mb4&interpolateParams=true&readTimeout=%s", cfg.User, cfg.Password, cfg.Host, cfg.Port, showStatusConnectionTimeout)
-	db, err := sql.Open("mysql", dbDSN)
+	if r.cfg.From.RawDBCfg != nil {
```
Comment (lance6716) [inline relay/relay.go:922]:
```suggestion
	if r.cfg.From.RawDBCfg == nil {
```

**Comment by lance6716** [comment]
failed Actions should be fixed by https://github.com/pingcap/dm/pull/1567, simply re-run

**Pull request merged**